### ISSUE TITLE
Error: Cannot read property 'close' of undefined

### ISSUE THREAD

**Issue by LaurensRietveld** [comunica/comunica#791]
#### Issue type:

- :bug: Bug 

____
#### Description:

For some queries, the error `Cannot read property 'close' of undefined` is thrown. 
To reproduce:

- `cd ./packages/actor-init-sparql/bin`
- Store the following query in `./query.sparql`
```
select ?x where {
    ?x <a:a> <b:b> .
    optional {}
    <c:c> <d:d> ?x.
}
```
- Run `node ./query.js http://fragments.dbpedia.org/2016-04/en -f ./query.sparql`

Some more information:
- Because this is a MWE, the optional clause is empty. The issue occurs for a populated optional clause as well
- Removing any of the three lines in the where clause works fine and wont throw an error

____
#### Environment:

`cd ./packages/actor-init-sparql/bin && node ./query.js -v`

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.19.2 (dev)
| node                | v12.20.1
| npm                 | 7.5.6
| yarn                | 1.22.5
| Operating System    | linux (Linux 5.8.0-43-generic)

#### Crash log:

https://pastebin.com/70kPU0vW

**Comment by rubensworks** [comment]
That's interesting. First time I see an error in the `NestedLoopJoin` implementation.

@joachimvh I suspect this is a problem in `asyncjoin`. Any idea what could be going wrong?

**Comment by joachimvh** [comment]
The error seems to imply that one of the asyncjoin parameters was `undefined` instead of a stream so the problem might be higher up. Unless it's expected that asyncjoin can handle undefined parameters.

**Comment by rubensworks** [comment]
Then our `bindingsStream` could be `undefined` here: https://github.com/comunica/comunica/blob/1c54ea3ce45ab00857d07b921c8ed263c10f0b8b/packages/actor-rdf-join-nestedloop/lib/ActorRdfJoinNestedLoop.ts#L22-L24
To be investigated...

**Comment by rubensworks** [comment]
@joachimvh I just did some debugging, and it looks like all params Comunica passes to are in fact properly defined.
The exact place where it crashes is here: https://github.com/joachimvh/asyncjoin/blob/master/join/NestedLoopJoin.js#L21
And more specifically, this happens at construction time:
* https://github.com/RubenVerborgh/AsyncIterator/blob/master/asynciterator.ts#L1015
* https://github.com/RubenVerborgh/AsyncIterator/blob/master/asynciterator.ts#L1045

I suspect wrapping `right.close()` into [`scheduleTask`](https://github.com/RubenVerborgh/AsyncIterator/blob/master/asynciterator.ts#L13) should fix the problem.

**Comment by joachimvh** [comment]
That does seem to fix it, will also have to change it in the other join classes.

Did not expect close to be called in the super constructor.

**Comment by joachimvh** [comment]
Should be fixed in asyncjoin v1.0.3

**Comment by rubensworks** [comment]
Seems to be working on my end as well, thanks!
